torchao/prototype/awq/example.py

```diff
@@ -0,0 +1,173 @@
+    return [cat_samples[:, i * block_size : (i + 1) * block_size] for i in range(n_samples)]
+
+def wiki2_eval(model, tokenizer, sequence_length):
+    testenc = load_dataset("wikitext", "wikitext-2-raw-v1", split="test")
+    testenc = tokenizer("\n\n".join(testenc["text"]), return_tensors="pt")
+    testenc = testenc.input_ids
+    nsamples = 100
+    model = model.eval()
+    # calculate perplexity
+    nlls = []
+    for i in tqdm(range(nsamples), desc="evaluating..."):
+        batch = testenc[:, i : i + sequence_length].to(
+            model.device
+        )
+        with torch.no_grad():
+            lm_logits = model(batch).logits
+        batch = batch.to("cpu")
+        lm_logits = lm_logits.to("cpu")
+        shift_logits = lm_logits[:, :-1, :].contiguous().float()
+        shift_labels = testenc[:, i : i + sequence_length][:, 1:].to("cpu")
+        loss_fct = torch.nn.CrossEntropyLoss()
+        loss = loss_fct(
+            shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1)
+        )
+        neg_log_likelihood = loss.float() * sequence_length
+        nlls.append(neg_log_likelihood)
+
+    ppl = torch.exp(torch.stack(nlls).sum() / (nsamples * sequence_length))
+    
+    return ppl
+
+def wikitext2_ppl(
+        repo_id: str,
+        quant: str, 
+        calibration_size: int =100, 
+        validation_size:int=100, 
+        group_size: int = 128, 
+        device="cuda", 
+        precision=torch.bfloat16, 
+        sequence_length=2048, 
+        compile=False,
+        model_save_path=None):
+    print(f"Loading model on {device}...")
+    torch.manual_seed(34)
+    t0 = time.time()
+    # load any model with torch.nn.linear layers
+    tokenizer = AutoTokenizer.from_pretrained(repo_id)
+    model = AutoModelForCausalLM.from_pretrained(repo_id, torch_dtype=precision).eval().to(device)
+    print(f"Time to load model: {time.time() - t0:.02f} seconds")
+
+    if quant.startswith("awq"):
+        quant_dtype = quant.split("-")[1]
+        quant_dtype = getattr(torch, quant_dtype, torch.bfloat16)
+        print(f"running {quant_dtype} calibration")
+        t0 = time.time()
+        
+        # insert observers to find average magnitude and calculate scales
+            
+        insert_awq_observer_(model,validation_size, sequence_length, quant_dtype=quant_dtype, group_size=group_size)
+        calibration_data = get_calib_dataset(tokenizer=tokenizer, n_samples=calibration_size, block_size=sequence_length)
+        for batch in calibration_data:
+            model(batch.to(device))
+            batch.to("cpu")
+        print(f"time for calibration: {time.time() - t0:.02f} seconds")
+
+        print(f"running {quant_dtype} quantization")
+        # use awq_uintx() to apply awq quantization
+        is_observed_linear = lambda m, fqn: isinstance(m, AWQObservedLinear)
+        
+        
+        t0 = time.time()
+        if "hqq" in quant:
+            from torchao.quantization.quant_primitives import (
```

Comment:
I guess generally might be useful to have a quant util function to host all these code..